File: src/texture/gradient_texture.cpp
/*******************************************************************************
    Copyright (c) The Taichi Authors (2016- ). All Rights Reserved.
    The use of this software is governed by the LICENSE file.
*******************************************************************************/

#include <taichi/visual/texture.h>

TC_NAMESPACE_BEGIN

class SphericalGradientTexture final : public Texture {
 private:
  Vector4 inside_val;
  Vector4 outside_val;
  real angle, sharpness;

 public:
  void initialize(const Config &config) override {
    Texture::initialize(config);
    TC_LOAD_CONFIG(inside_val, Vector4(0.0_f));
    TC_LOAD_CONFIG(outside_val, Vector4(1.0_f));
    TC_LOAD_CONFIG(angle, 30.0_f);
    TC_LOAD_CONFIG(sharpness, 1.0_f);
  }

  virtual Vector4 sample(const Vector3 &coord) const override {
    real theta_d = (coord.x * 2 + 1) * pi, phi_d = (coord.y - 0.5f) * pi;
    real angle_ =
        std::acos(std::sin(theta_d) * std::cos(phi_d)) * (180.0_f / pi);
    real t = std::tanh((angle_ - angle) * sharpness * 0.01f) * 0.5f + 0.5f;
    return lerp(t, inside_val, outside_val);
  }
};

TC_IMPLEMENTATION(Texture, SphericalGradientTexture, "spherical_gradient");

TC_NAMESPACE_END

File: projects/examples/rendering/trashbin/gradient_texture.py
import math
from taichi.misc.util import Vector
from taichi.visual import *
from taichi.visual.post_process import *
from taichi.visual.texture import Texture
import time

if __name__ == '__main__':
  while True:
    for i in range(100):
      envmap_texture = Texture(
          'spherical_gradient',
          inside_val=(1, 1, 10, 1),
          outside_val=(0.5, 0.5, 0.5, 1),
          angle=3,
          sharpness=5)
      envmap_texture.show(res=(512, 512), post_processor=LDRDisplay())
